framework/util/page_guard_manager.cpp

```diff
@@ -199,6 +209,377 @@ void PageGuardManager::InitializeSystemExceptionContext(void)
+            }
+            else
+            {
+                page_offset -= memory_info->aligned_offset;
+            }
+
+            assert(memory_info->shadow_memory != nullptr);
+            uint8_t* destination_address = static_cast<uint8_t*>(memory_info->shadow_memory) + page_offset;
+            uint8_t* source_address      = static_cast<uint8_t*>(memory_info->mapped_memory) + page_offset;
+
+            struct uffdio_copy copy;
+            copy.dst  = reinterpret_cast<uint64_t>(destination_address);
+            copy.src  = reinterpret_cast<uint64_t>(source_address);
+            copy.len  = system_page_size_;
+            copy.mode = 0;
+
+            assert(uffd_fd != -1);
+            if (ioctl(uffd_fd, UFFDIO_COPY, &copy))
+            {
+                GFXRECON_LOG_ERROR("ioctl/uffdio_copy: %s", strerror(errno));
+            }
+
+            if (copy.copy != system_page_size_)
+            {
+                GFXRECON_LOG_ERROR("Unexpected copy.copy\n");
+            }
+        }
+    }
+}
+
+void* PageGuardManager::UffdHandlerThread(void* args)
+{
+    GFXRECON_UNREFERENCED_PARAMETER(args);
+    assert(uffd_fd_ != -1);
+
+    for (;;)
+    {
+        struct pollfd pollfd[1];
+        pollfd[0].fd     = uffd_fd_;
+        pollfd[0].events = POLLIN | POLLERR | POLLPRI;
+
+        // wait for a userfaultfd event to occur
+        int pollres = poll(pollfd, 1, 200);
+
+        if (uffd_stop_handler_thread.load())
+        {
+            break;
+        }
+
+        switch (pollres)
+        {
+            case -1:
+                GFXRECON_LOG_WARNING("poll/userfaultfd: %s", strerror(errno));
+                continue;
+            case 0:
+                continue;
+            case 1:
+                break;
+            default:
+                GFXRECON_LOG_WARNING("unexpected poll result: %d", pollres);
+                continue;
+        }
+
+        if (pollfd[0].revents & POLLERR)
+        {
+            GFXRECON_LOG_ERROR("pollerr: %s", strerror(errno));
+            continue;
+        }
+
+        if ((pollfd[0].revents & POLLIN) != POLLIN)
+        {
+            continue;
+        }
+
+        static struct uffd_msg msg[256];
+        int                    readres = read(uffd_fd_, &msg, sizeof(msg));
+        if (readres == -1)
+        {
+            if (errno == EAGAIN)
+            {
+                GFXRECON_LOG_ERROR("read/userfaultfd EAGAIN: %s", strerror(errno));
+                continue;
+            }
+
+            GFXRECON_LOG_ERROR("read/userfaultfd: %s", strerror(errno));
+            break;
+        }
+
+        if (readres % sizeof(uffd_msg))
+        {
+            GFXRECON_LOG_ERROR("Unexpected read size\n");
+            break;
+        }
+
+        for (int i = 0; i < readres / sizeof(struct uffd_msg); ++i)
+        {
+            if (msg[i].event == UFFD_EVENT_PAGEFAULT)
+            {
+                UffdHandleFault(uffd_fd_, msg[i].arg.pagefault.address, msg[i].arg.pagefault.flags);
```

Comment:
I think if you handle multiple events here, you need to use the DONTWAKE flags in the mode field for all but the latest event, otherwise you will get more race conditions, as the calling code wakes up and can do reads/writes that we presumably cannot catch while still processing more events.

Reply:
Ah, good catch.